src/udp_address.cpp

```diff
@@ -129,88 +123,55 @@ int zmq::udp_address_t::resolve (const char *name_, bool bind_)
             return -1;
         }
 
-        interface_addr.set_port (port);
-
-        dest_address = target_addr.ipv4;
-        bind_address = interface_addr.ipv4;
+        bind_address.set_port (port);
     } else {
         //  If we don't have an explicit interface specifier then the URL is
         //  ambiguous: if the target address is multicast then it's the
         //  destination address and the bind address is ANY, if it's unicast
         //  then it's the bind address when 'bind_' is true and the destination
         //  otherwise
-        ip_addr_t any = ip_addr_t::any (family);
-        any.set_port (port);
-
-        if (is_multicast) {
-            dest_address = target_addr.ipv4;
-            bind_address = any.ipv4;
+        if (is_multicast || !bind_) {
+            bind_address = ip_addr_t::any (target_address.family ());
+            bind_address.set_port (port);
         } else {
-            if (bind_) {
-                dest_address = target_addr.ipv4;
-                bind_address = target_addr.ipv4;
-            } else {
-                dest_address = target_addr.ipv4;
-                bind_address = any.ipv4;
-            }
+            //  If we were asked for a bind socket and the address
+            //  provided was not multicast then it was really meant as
+            //  a bind address and the target_address is useless.
+            bind_address = target_address;
         }
     }
 
-    if (is_multicast) {
-        multicast = dest_address.sin_addr;
+    if (bind_address.family () != target_address.family ()) {
+        errno = EINVAL;
+        return -1;
     }
 
-    address = name_;
-
     return 0;
 }
 
-int zmq::udp_address_t::to_string (std::string &addr_)
+int zmq::udp_address_t::family () const
 {
-    addr_ = address;
-    return 0;
+    return bind_address.family ();
 }
 
 bool zmq::udp_address_t::is_mcast () const
 {
     return is_multicast;
 }
 
-const sockaddr *zmq::udp_address_t::bind_addr () const
-{
-    return (sockaddr *) &bind_address;
-}
-
-socklen_t zmq::udp_address_t::bind_addrlen () const
-{
-    return sizeof (sockaddr_in);
-}
-
-const sockaddr *zmq::udp_address_t::dest_addr () const
-{
-    return (sockaddr *) &dest_address;
-}
-
-socklen_t zmq::udp_address_t::dest_addrlen () const
+const zmq::ip_addr_t *zmq::udp_address_t::bind_addr () const
 {
-    return sizeof (sockaddr_in);
+    return &bind_address;
 }
 
-const in_addr zmq::udp_address_t::multicast_ip () const
+const zmq::ip_addr_t *zmq::udp_address_t::target_addr () const
 {
-    return multicast;
+    return &target_address;
 }
 
-const in_addr zmq::udp_address_t::interface_ip () const
-{
-    return iface;
-}
-
-#if defined ZMQ_HAVE_WINDOWS
-unsigned short zmq::udp_address_t::family () const
-#else
-sa_family_t zmq::udp_address_t::family () const
-#endif
+int zmq::udp_address_t::to_string (std::string &addr_)
 {
-    return AF_INET;
+    // XXX what do (factor TCP code?)
+    addr_ = address;
+    return 0;
```

Comment:
What exactly are the requirements for this `to_string` method? The current implementation just returns the string passed to the resolver but the TCP implementation actually bothers re-generating the string from the sockaddr values.

Reply:
I have a feeling there was something related to disconnect/unbind - check the git log, there might be some clues